src/libsync/propagatecommonzsync.cpp

```diff
@@ -0,0 +1,343 @@
+        return false;
+    }
+    if (item->_remotePerm.hasPermission(RemotePermissions::IsMounted) ||
+        item->_remotePerm.hasPermission(RemotePermissions::IsMountedSub)) {
+        qCInfo(lcPropagator) << "[zsync disabled] External storage not supported.";
+        return false;
+    }
+    if (!propagator->syncOptions()._deltaSyncEnabled) {
+        qCInfo(lcPropagator) << "[zsync disabled] Client configuration option.";
+        return false;
+    }
+    if (item->_size < propagator->syncOptions()._deltaSyncMinFileSize) {
+        qCInfo(lcPropagator) << "[zsync disabled] File size is smaller than minimum.";
+        return false;
+    }
+
+    return true;
+}
+
+QUrl zsyncMetadataUrl(OwncloudPropagator *propagator, QString &path)
+{
+    return Utility::concatUrlPath(propagator->account()->davUrl(), propagator->_remoteFolder + path,
+                                  QList<QPair<QString, QString>> ({{"zsync", nullptr}}));
+}
+
+void ZsyncSeedRunnable::run()
+{
+    // Create a temporary file to use with zsync_begin()
+    QTemporaryFile zsyncControlFile;
+    zsyncControlFile.open();
+    zsyncControlFile.write(_zsyncData.constData(), _zsyncData.size());
+    zsyncControlFile.flush();
+
+    int fileHandle = zsyncControlFile.handle();
+    FILE* f = fdopen(dup(fileHandle), "r");
+    zsyncControlFile.close();
+    rewind(f);
+
+    zsync_unique_ptr<struct zsync_state> zs(zsync_parse(f), [](struct zsync_state* zs) {
+        zsync_end(zs);
+    });
+    if (!zs) {
+        QString errorString = tr("Unable to parse zsync file.");
+        fclose(f);
+        emit failedSignal(errorString);
+        return;
+    }
+
+    std::string tmp_file;
+    if (!_tmpFilePath.isEmpty()) {
+        tmp_file = _tmpFilePath.toStdString();
+    } else {
+        QTemporaryFile tmpFile;
+        tmpFile.open();
+        tmp_file = tmpFile.fileName().toStdString();
+        tmpFile.close();
+    }
+
+    const char * tfname = tmp_file.c_str();
+    if (zsync_rename_file(zs.get(), tfname) != 0) {
+        QString errorString = tr("Unable to rename temporary file.");
+        emit failedSignal(errorString);
+        return;
+    }
+
+    if (zsync_begin(zs.get(), f)) {
+        QString errorString = tr("Unable to begin zsync.");
+        emit failedSignal(errorString);
+        return;
+    }
+
+    fclose(f);
+
+    std::string tmp = _zsyncFilePath.toStdString();
+    const char * fname = tmp.c_str();
+    {
+        /* Simple uncompressed file - open it */
+        QFile file(fname);
+        if (!file.open(QIODevice::ReadOnly)) {
+            QString errorString = tr("Unable to open file.");
+            emit failedSignal(errorString);
+            return;
+        }
+
+        /* Give the contents to libzsync to read, to find any content that
+         * is part of the target file. */
+        qCInfo(lcZsyncSeed) << "Reading seed file:" << _zsyncFilePath;
+        int fileHandle = file.handle();
+        FILE* f = fdopen(dup(fileHandle), "r");
```

Comment:
Duplicating the handle was because I read that closing a file handle for a QFile is a bug: https://stackoverflow.com/questions/9465727/convert-qfile-to-file/16876527#16876527